pex/requirements.py

```diff
@@ -5,67 +5,586 @@
+    if fname.endswith(".whl"):
+        project_name, _ = fname.split("-", 1)
+        return project_name
+
+    # Handle sdists:
+    #
+    # The sdist name format is specified here:
+    #   https://www.python.org/dev/peps/pep-0625/#specification.
+    # We allow a few more legacy extensions.
+    if fname.endswith((".tar.gz", ".zip")):
+        project_name, _ = fname.rsplit("-", 1)
+        return project_name
+
+    return None
+
+
+def _try_parse_pip_local_project_marker(path):
+    # type: (str) -> Tuple[str, Optional[Marker]]
+    project_requirement = os.path.basename(path)
+    try:
+        req = Requirement.parse(project_requirement)
+        stripped_path = os.path.join(os.path.dirname(path), req.name)
+        return stripped_path, req.marker
+    except (RequirementParseError, ValueError):
+        return path, None
+
+
+def _split_direct_references(processed_text):
+    # type: (str) -> Tuple[str, Optional[str]]
+    parts = processed_text.split("@", 1)
+    if len(parts) == 1:
+        return processed_text, None
+    return parts[0].strip(), parts[1].strip()
+
+
+def _parse_requirement_line(
+    line,  # type: LogicalLine
+    basepath=None,  # type: Optional[str]
+):
+    # type: (...) -> ReqInfo
+
+    basepath = basepath or os.getcwd()
+
+    editable, processed_text = _strip_requirement_options(line)
+
+    # Handle urls (Pip proprietary).
+    parsed_url = urlparse.urlparse(processed_text)
+    if _is_recognized_pip_url_scheme(parsed_url.scheme):
+        project_name, marker = _try_parse_fragment_project_name_and_marker(parsed_url.fragment)
+        if not project_name:
+            project_name = _try_parse_project_name_from_path(parsed_url.path)
+        url = parsed_url._replace(fragment="").geturl()
+        return ReqInfo.create(
+            line, project_name=project_name, url=url, marker=marker, editable=editable
+        )
+
+    # Handle local archives and project directories (Pip proprietary).
+    maybe_path, marker = _try_parse_pip_local_project_marker(processed_text)
+    maybe_abs_path = maybe_path if os.path.isabs(maybe_path) else os.path.join(basepath, maybe_path)
+    if any(
+        os.path.isfile(os.path.join(maybe_abs_path, *p))
+        for p in ((), ("setup.py",), ("pyproject.toml",))
+    ):
+        archive_or_project_path = os.path.realpath(maybe_abs_path)
+        is_local_project = os.path.isdir(archive_or_project_path)
+        project_name = (
+            None if is_local_project else _try_parse_project_name_from_path(archive_or_project_path)
+        )
+        return ReqInfo.create(
+            line,
+            project_name=project_name,
+            url=archive_or_project_path,
+            marker=marker,
+            editable=editable,
+            is_local_project=is_local_project,
+        )
+
+    # Handle PEP-440. See: https://www.python.org/dev/peps/pep-0440.
+    #
+    # The `pkg_resources.Requirement.parse` method does all of this for us (via
```

Comment:
N.B.: Although Pex vendors the latest packaging which does handle direct references, It currently does not import correctly in tests, failing on `pex.thirdparty.packaging.six` for inscrutable reasons. I cut losses and used `pkg_resources.Requirement.parse` from our old vendored version of setuptools since the `_split_direct_references` is fairly simple.